File: solutions/python/01-jq3/code/app/main.py
import sys


class LRUCache:
    def __init__(self, capacity: int):
        """Initialize LRU Cache with given capacity"""
        self.capacity = capacity
        self.cache = {}
    
    def get(self, key: str) -> str | None:
        """Get value by key, return None if not found"""
        return self.cache.get(key)
    
    def put(self, key: str, value: str) -> None:
        """Insert or update key-value pair"""
        self.cache[key] = value
    
    def size(self) -> int:
        """Return the number of items in the cache"""
        return len(self.cache)


def main():
    cache = None
    
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        
        parts = line.split()
        command = parts[0]
        
        if command == "INIT":
            capacity = int(parts[1])
            cache = LRUCache(capacity)
            print("OK")
        
        elif command == "PUT":
            if cache is None:
                print("ERROR: Cache not initialized")
                continue
            key = parts[1]
            value = parts[2]
            cache.put(key, value)
            print("OK")
        
        elif command == "GET":
            if cache is None:
                print("ERROR: Cache not initialized")
                continue
            key = parts[1]
            result = cache.get(key)
            if result is None:
                print("NULL")
            else:
                print(result)
        
        elif command == "SIZE":
            if cache is None:
                print("ERROR: Cache not initialized")
                continue
            print(cache.size())
        
        else:
            print(f"ERROR: Unknown command: {command}")


if __name__ == "__main__":
    main()

File: solutions/java/01-jq3/code/src/main/java/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

/**
 * LRU Cache implementation
 */
class LRUCache {
    private final int capacity;
    private final Map<String, String> cache;
    
    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>();
    }
    
    public String get(String key) {
        return cache.get(key);
    }
    
    public void put(String key, String value) {
        cache.put(key, value);
    }
    
    public int size() {
        return cache.size();
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        LRUCache cache = null;
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String line;
        
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) {
                continue;
            }
            
            String[] parts = line.split("\\s+");
            String command = parts[0];
            
            switch (command) {
                case "INIT":
                    int capacity = Integer.parseInt(parts[1]);
                    cache = new LRUCache(capacity);
                    System.out.println("OK");
                    break;
                    
                case "PUT":
                    if (cache == null) {
                        System.out.println("ERROR: Cache not initialized");
                        continue;
                    }
                    String putKey = parts[1];
                    String putValue = parts[2];
                    cache.put(putKey, putValue);
                    System.out.println("OK");
                    break;
                    
                case "GET":
                    if (cache == null) {
                        System.out.println("ERROR: Cache not initialized");
                        continue;
                    }
                    String getKey = parts[1];
                    String result = cache.get(getKey);
                    if (result == null) {
                        System.out.println("NULL");
                    } else {
                        System.out.println(result);
                    }
                    break;
                    
                case "SIZE":
                    if (cache == null) {
                        System.out.println("ERROR: Cache not initialized");
                        continue;
                    }
                    System.out.println(cache.size());
                    break;
                    
                default:
                    System.out.println("ERROR: Unknown command: " + command);
                    break;
            }
        }
    }
}
